gpytorch/models/exact_prediction_strategies.py

```diff
@@ -35,9 +36,14 @@ def prediction_strategy(train_inputs, train_prior_dist, train_labels, likelihood
 
 class DefaultPredictionStrategy(object):
     def __init__(self, train_inputs, train_prior_dist, train_labels, likelihood, root=None, inv_root=None):
+        # Get training shape
+        self._train_shape = train_prior_dist.event_shape
+        # Ensure that the event_shape actually matches the number of inputs
+        if self._train_shape[0] != train_inputs[0].size(-2):
+            self._train_shape = torch.Size([train_inputs[0].size(-2), *self._train_shape[1:]])
+
         # Flatten the training labels
-        train_shape = train_prior_dist.event_shape
-        train_labels = train_labels.reshape(*train_labels.shape[: -len(train_shape)], train_shape.numel())
+        train_labels = train_labels.reshape(*train_labels.shape[: -len(self.train_shape)], self.num_train)
```

Comment:
@wjmaddox I believe this change is what breaks the `HigherOrderGP` model (since that model takes multi-dimensional targets and flattens them internally).

Reply:
@gpleiss is this change necessary for the functionality of the PR or is this just some auxiliary cleanup?

Reply:
It was necessary. Let me take a look, and I'll see if we can't come to some better solution

Reply:
The issue is these lines of code in `MultivariateNormal` (https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/distributions/multivariate_normal.py#L45)

```python
            if not isinstance(self._covar, RootLazyTensor):
                should_have_separate_event = False
            else:
                should_have_separate_event = True
            if should_have_separate_event:
                should_have_separate_event = self._covar.root.shape[-1] > self._covar.root.shape[-2]
            if not should_have_separate_event:
                event_shape = self.loc.shape[-1:]
            else:
                event_shape = self._covar.root.shape[-1:]
```

For RFF models, it is possible that the number of random features exceeds the amount of data (this can still be advantageous for sampling purposes). In that case, if `N=50` and `D=100` (the number of features), then the MVN will have an event shape of `[100]`. That leads to this undesirable behavior in ExactPredictionStrategies (old version).

```python
        train_shape = train_prior_dist.event_shape
        train_labels = train_labels.reshape(*train_labels.shape[: -len(train_shape)], train_shape.numel())
```

Here, `train_labels` is of size `N=50`, but we're trying to reshape it to size `train_shape=[100]`.

Those lines of code in MVN were added in #1355 . It seemed benign at the time when I reviewed it, but now it seems like it might cause problems. @wjmaddox and @Balandat - what's the best way to resolve these issues?

Reply:
The proper way to resolve this would probably be to use torch's `LowRankMultivariateNormal`, but that would be quite a bit of work. The full-on complete solution would be to use `LinearOperator` with the `__torch_function__` tensor interface and just have the torch code do everything but properly dispatch to the lazy tensor methods when needed. Even more work. 

A realistic fix in the short term is probably to do something like introducing an additional `base_sample_shape` parameter. Then `event_shape` would be the size of a single sample, and `base_sample_shape` would be the size of a single base sample to construct one sample. These would agree except for the low-rank case.

Reply:
Ugh, I'd thought that allowing MC samples to be drawn botorch style would fix all of the `D > N` issues and not create new ones. I think a `base_sample_shape` parameter is probably the cleanest way out as it should resolve both sets of issues.

Reply:
I'll go ahead and try to set that up now.